Description: Internal setting for hide popups overlapping other controls
 Create an empty "nopopupdropdown" file in the "tdata" directory to activate
 the changes: `touch ~/.local/share/TelegramDesktop/tdata/nopopupdropdown`.
 There is a cheat code of the same name on the Settings page.
Author: Nicholas Guriev <guriev-ns@ya.ru>
Forwarded: https://github.com/telegramdesktop/tdesktop/pull/8009
Last-Update: Sat, 27 Feb 2021 09:51:39 +0300

--- a/Telegram/SourceFiles/chat_helpers/tabbed_panel.cpp
+++ b/Telegram/SourceFiles/chat_helpers/tabbed_panel.cpp
@@ -231,7 +231,9 @@ void TabbedPanel::leaveEventHook(QEvent
 }
 
 void TabbedPanel::otherEnter() {
-	showAnimated();
+	if (cPopupDropdown()) {
+		showAnimated();
+	}
 }
 
 void TabbedPanel::otherLeave() {
--- a/Telegram/SourceFiles/core/application.cpp
+++ b/Telegram/SourceFiles/core/application.cpp
@@ -213,6 +213,7 @@ void Application::run() {
 		App::quit();
 		return;
 	}
+	cSetPopupDropdown(!QFileInfo(cWorkingDir() + u"tdata/nopopupdropdown"_q).isFile());
 
 	_translator = std::make_unique<Lang::Translator>();
 	QCoreApplication::instance()->installTranslator(_translator.get());
--- a/Telegram/SourceFiles/history/history_widget.cpp
+++ b/Telegram/SourceFiles/history/history_widget.cpp
@@ -842,7 +842,11 @@ void HistoryWidget::initTabbedSelector()
 	refreshTabbedPanel();
 
 	_tabbedSelectorToggle->addClickHandler([=] {
-		toggleTabbedSelectorMode();
+		if (!cPopupDropdown() && _tabbedPanel && _tabbedPanel->isHidden()) {
+			_tabbedPanel->showAnimated();
+		} else {
+			toggleTabbedSelectorMode();
+		}
 	});
 
 	const auto selector = controller()->tabbedSelector();
--- a/Telegram/SourceFiles/history/view/controls/history_view_compose_controls.cpp
+++ b/Telegram/SourceFiles/history/view/controls/history_view_compose_controls.cpp
@@ -1490,7 +1490,11 @@ void ComposeControls::initTabbedSelector
 	}
 
 	_tabbedSelectorToggle->addClickHandler([=] {
-		toggleTabbedSelectorMode();
+		if (!cPopupDropdown() && _tabbedPanel && _tabbedPanel->isHidden()) {
+			_tabbedPanel->showAnimated();
+		} else {
+			toggleTabbedSelectorMode();
+		}
 	});
 
 	const auto selector = _window->tabbedSelector();
--- a/Telegram/SourceFiles/media/view/media_view_overlay_widget.cpp
+++ b/Telegram/SourceFiles/media/view/media_view_overlay_widget.cpp
@@ -1401,7 +1401,7 @@ void OverlayWidget::dropdownHidden() {
 	_ignoringDropdown = true;
 	_lastMouseMovePos = _widget->mapFromGlobal(QCursor::pos());
 	updateOver(_lastMouseMovePos);
-	_ignoringDropdown = false;
+	_ignoringDropdown = !cPopupDropdown();
 	if (!_controlsHideTimer.isActive()) {
 		hideControls(true);
 	}
@@ -2201,6 +2201,7 @@ void OverlayWidget::clearControlsState()
 	if (!_animationOpacities.empty()) {
 		_animationOpacities.clear();
 	}
+	_ignoringDropdown = !cPopupDropdown();
 }
 
 not_null<QWindow*> OverlayWidget::window() const {
--- a/Telegram/SourceFiles/settings.cpp
+++ b/Telegram/SourceFiles/settings.cpp
@@ -63,6 +63,7 @@ float64 gRetinaFactor = 1.;
 int32 gIntRetinaFactor = 1;
 
 int gOtherOnline = 0;
+bool gPopupDropdown = true;
 
 int32 gAutoDownloadPhoto = 0; // all auto download
 int32 gAutoDownloadAudio = 0;
--- a/Telegram/SourceFiles/settings.h
+++ b/Telegram/SourceFiles/settings.h
@@ -108,6 +108,7 @@ DeclareSetting(QStringList, SendPaths);
 DeclareSetting(QString, StartUrl);
 
 DeclareSetting(int, OtherOnline);
+DeclareSetting(bool, PopupDropdown);
 
 inline void cChangeTimeFormat(const QString &newFormat) {
 	if (!newFormat.isEmpty()) cSetTimeFormat(newFormat);
--- a/Telegram/SourceFiles/settings/settings_codes.cpp
+++ b/Telegram/SourceFiles/settings/settings_codes.cpp
@@ -140,6 +140,17 @@ auto GenerateCodes() {
 		Core::Application::RegisterUrlScheme();
 		Ui::Toast::Show("Forced custom scheme register.");
 	});
+	codes.emplace(qsl("nopopupdropdown"), [](SessionController *window) {
+		if (cPopupDropdown()) {
+			cSetPopupDropdown(false);
+			QFile(cWorkingDir() + u"tdata/nopopupdropdown"_q).open(QIODevice::WriteOnly);
+			Ui::Toast::Show("PopupDropdown = FALSE");
+		} else {
+			cSetPopupDropdown(true);
+			QFile(cWorkingDir() + u"tdata/nopopupdropdown"_q).remove();
+			Ui::Toast::Show("PopupDropdown = TRUE");
+		}
+	});
 
 #if defined Q_OS_WIN || defined Q_OS_MAC
 	codes.emplace(qsl("freetype"), [](SessionController *window) {
--- a/Telegram/SourceFiles/ui/filter_icon_panel.cpp
+++ b/Telegram/SourceFiles/ui/filter_icon_panel.cpp
@@ -282,7 +282,9 @@ void FilterIconPanel::leaveEventHook(QEv
 }
 
 void FilterIconPanel::otherEnter() {
-	showAnimated();
+	if (cPopupDropdown()) {
+		showAnimated();
+	}
 }
 
 void FilterIconPanel::otherLeave() {
